src/MCPServer/lib/workflow.py

```diff
@@ -0,0 +1,315 @@
+        """Replace status labels with their IDs.
+
+        In JSON, a job status is encoded using its English label, e.g. "Failed"
+        instead of the corresponding value in ``JOB.STATUS_FAILED``. This
+        method decodes the statuses so it becomes easier to work with them
+        internally.
+        """
+        self._src["fallback_job_status"] = \
+            _STATUSES[self._src["fallback_job_status"]]
+        for obj in self._src["exit_codes"].values():
+            obj["job_status"] = _STATUSES[obj["job_status"]]
+
+    def _decode_translations(self):
+        self._src["description"] = self._decode_translation(
+            self._src["description"])
+        self._src["group"] = self._decode_translation(
+            self._src["group"])
+        config = self._src["config"]
+        if config["@manager"] == "linkTaskManagerReplacementDicFromChoice":
+            for item in config["replacements"]:
+                item["description"] = self._decode_translation(
+                    item["description"])
+
+    @property
+    def manager(self):
+        name = self._src["config"]["@manager"]
+        module = importlib.import_module(name)
+        return getattr(module, name)
+
+    @property
+    def config(self):
+        return self._src["config"]
+
+    def get_next_link(self, code):
+        code = text_type(code)
+        try:
+            link_id = self._src["exit_codes"][code]["link_id"]
+        except KeyError:
+            link_id = self._src["fallback_link_id"]
+        return self._workflow.get_link(link_id)
+
+    def get_status_id(self, code):
+        """Return the expected Job status ID given an exit code."""
+        code = text_type(code)
+        try:
+            status_id = self._src["exit_codes"][code]["job_status"]
+        except KeyError:
+            status_id = self._src["fallback_job_status"]
+        return status_id
+
+
+class WatchedDir(BaseLink):
+    def __init__(self, attrs, workflow):
+        self.path = attrs["path"]
+        self._src = attrs
+        self._workflow = workflow
+
+    def __str__(self):
+        return self.path
+
+    def __repr__(self):
+        return "Watched directory <{}>".format(self.path)
+
+    def __getitem__(self, key):
+        return self._src[key]
+
+    @property
+    def chain(self):
+        return self._workflow.get_chain(self._src["chain_id"])
+
+
+class WorkflowJSONDecoder(json.JSONDecoder):
+    def decode(self, foo, **kwargs):
+        parsed_json = super(WorkflowJSONDecoder, self).decode(foo, **kwargs)
+        return Workflow(parsed_json)
+
+
+def load(fp):
+    """Read JSON document from file-like object, validate and decode it."""
+    blob = fp.read()  # Read once, used twice.
+    _validate(blob)
+    return json.loads(blob, cls=WorkflowJSONDecoder)
+
+
+def _validate(blob):
+    """Decode and validate the JSON document."""
+    validate(
+        json.loads(blob), _get_schema(),
+        format_checker=FormatChecker())
```

Comment:
I haven't done extensive testing of the schema validation, but it looks good (see below). Only question, do we want to capture the `ValidationError` and handle the output more gracefully if we have the opportunity?:

```
archivematica-mcp-server_1       | ERROR     2019-01-01 19:11:01  archivematica.mcp.server.rpcserver:workflow:load:308:  Validating
archivematica-mcp-server_1       | ERROR     2019-01-01 19:11:01  archivematica.mcp.server:archivematicaMCP:_except_hook_log_everything:352:  Uncaught exception
archivematica-mcp-server_1       | Traceback (most recent call last):
archivematica-mcp-server_1       |   File "/src/MCPServer/lib/archivematicaMCP.py", line 370, in <module>
archivematica-mcp-server_1       |     workflow = load_workflow(fp)
archivematica-mcp-server_1       |   File "/src/MCPServer/lib/workflow.py", line 309, in load
archivematica-mcp-server_1       |     _validate(blob)
archivematica-mcp-server_1       |   File "/src/MCPServer/lib/workflow.py", line 318, in _validate
archivematica-mcp-server_1       |     format_checker=FormatChecker())
archivematica-mcp-server_1       |   File "/usr/local/lib/python2.7/site-packages/jsonschema/validators.py", line 541, in validate
archivematica-mcp-server_1       |     cls(schema, *args, **kwargs).validate(instance)
archivematica-mcp-server_1       |   File "/usr/local/lib/python2.7/site-packages/jsonschema/validators.py", line 130, in validate
archivematica-mcp-server_1       |     raise error
archivematica-mcp-server_1       | ValidationError: Additional properties are not allowed (u'config_x' was unexpected)
archivematica-mcp-server_1       | 
archivematica-mcp-server_1       | Failed validating u'additionalProperties' in schema[u'properties'][u'links'][u'additionalProperties']:
archivematica-mcp-server_1       |     {u'@comment': u'`link` is synonymous with MicroServiceChainLink and its related attributes including task manager configurations.',
archivematica-mcp-server_1       |      u'additionalProperties': False,
archivematica-mcp-server_1       |      u'properties': {u'config': {u'oneOf': [{u'$ref': u'#/definitions/link_model_StandardTaskFile'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_StandardTaskDir'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_ReplacementDic'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_ChainChoice'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_GetServiceOutput'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_GetServiceOutputPrompt'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_GetUnitVar'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_SetUnitVar'}]},
archivematica-mcp-server_1       |                      u'description': {u'$ref': u'#/definitions/translations'},
archivematica-mcp-server_1       |                      u'exit_codes': {u'additionalProperties': {u'properties': {u'job_status': {u'oneOf': [{u'$ref': u'#/definitions/job_status'},
archivematica-mcp-server_1       |                                                                                                           {u'type': u'null'}]},
archivematica-mcp-server_1       |                                                                                u'link_id': {u'oneOf': [{u'$ref': u'#/definitions/uuid'},
archivematica-mcp-server_1       |                                                                                                        {u'type': u'null'}]}},
archivematica-mcp-server_1       |                                                                u'type': u'object'},
archivematica-mcp-server_1       |                                      u'minProperties': 0,
archivematica-mcp-server_1       |                                      u'type': u'object'},
archivematica-mcp-server_1       |                      u'fallback_job_status': {u'oneOf': [{u'$ref': u'#/definitions/job_status'},
archivematica-mcp-server_1       |                                                          {u'type': u'null'}]},
archivematica-mcp-server_1       |                      u'fallback_link_id': {u'oneOf': [{u'$ref': u'#/definitions/uuid'},
archivematica-mcp-server_1       |                                                       {u'type': u'null'}]},
archivematica-mcp-server_1       |                      u'group': {u'$ref': u'#/definitions/translations'}},
archivematica-mcp-server_1       |      u'required': [u'config',
archivematica-mcp-server_1       |                    u'exit_codes',
archivematica-mcp-server_1       |                    u'fallback_job_status',
archivematica-mcp-server_1       |                    u'fallback_link_id',
archivematica-mcp-server_1       |                    u'description',
archivematica-mcp-server_1       |                    u'group'],
archivematica-mcp-server_1       |      u'type': u'object'}
archivematica-mcp-server_1       | 
archivematica-mcp-server_1       | On instance[u'links'][u'c2e6600d-cd26-42ed-bed5-95d41c06e37b']:
archivematica-mcp-server_1       |     {u'config_x': {u'@manager': u'linkTaskManagerDirectories',
archivematica-mcp-server_1       |                    u'@model': u'StandardTaskconfig_x',
archivematica-mcp-server_1       |                    u'arguments': u'"%SIPDirectory%" "%sharedPath%watchedDirectories/approveNormalization/." "%SIPUUID%" "%sharedPath%" "%SIPUUID%" "%sharedPath%"',
archivematica-mcp-server_1       |                    u'execute': u'moveSIP_v0.0',
archivematica-mcp-server_1       |                    u'filter_file_end': None,
archivematica-mcp-server_1       |                    u'filter_file_start': None,
archivematica-mcp-server_1       |                    u'filter_subdir': None,
archivematica-mcp-server_1       |                    u'stderr_file': None,
archivematica-mcp-server_1       |                    u'stdout_file': None},
archivematica-mcp-server_1       |      u'description': {u'en': u'Move to approve normalization directory',
archivematica-mcp-server_1       |                       u'es': u'Move to approve normalization directory',
archivematica-mcp-server_1       |                       u'fr': u'Move to approve normalization directory',
archivematica-mcp-server_1       |                       u'ja': u'Move to approve normalization directory',
archivematica-mcp-server_1       |                       u'pt_BR': u'Mover para aprovar o diret\xf3rio de normaliza\xe7\xe3o',
archivematica-mcp-server_1       |                       u'sv': u'Flytta till godk\xe4nn normaliserings-katalog'},
archivematica-mcp-server_1       |      u'exit_codes': {u'0': {u'job_status': u'Completed successfully',
archivematica-mcp-server_1       |                             u'link_id': None}},
archivematica-mcp-server_1       |      u'fallback_job_status': u'Failed',
archivematica-mcp-server_1       |      u'fallback_link_id': None,
archivematica-mcp-server_1       |      u'group': {u'en': u'Normalize',
archivematica-mcp-server_1       |                 u'es': u'Normalizar',
archivematica-mcp-server_1       |                 u'fr': u'Normaliser ',
archivematica-mcp-server_1       |                 u'ja': u'Normalize',
archivematica-mcp-server_1       |                 u'pt_BR': u'Normalizar',
archivematica-mcp-server_1       |                 u'sv': u'Normalisera'}}
archivematica-mcp-server_1       | Traceback (most recent call last):
archivematica-mcp-server_1       |   File "/src/MCPServer/lib/archivematicaMCP.py", line 370, in <module>
archivematica-mcp-server_1       |     workflow = load_workflow(fp)
archivematica-mcp-server_1       |   File "/src/MCPServer/lib/workflow.py", line 309, in load
archivematica-mcp-server_1       |     _validate(blob)
archivematica-mcp-server_1       |   File "/src/MCPServer/lib/workflow.py", line 318, in _validate
archivematica-mcp-server_1       |     format_checker=FormatChecker())
archivematica-mcp-server_1       |   File "/usr/local/lib/python2.7/site-packages/jsonschema/validators.py", line 541, in validate
archivematica-mcp-server_1       |     cls(schema, *args, **kwargs).validate(instance)
archivematica-mcp-server_1       |   File "/usr/local/lib/python2.7/site-packages/jsonschema/validators.py", line 130, in validate
archivematica-mcp-server_1       |     raise error
archivematica-mcp-server_1       | jsonschema.exceptions.ValidationError: Additional properties are not allowed (u'config_x' was unexpected)
archivematica-mcp-server_1       | 
archivematica-mcp-server_1       | Failed validating u'additionalProperties' in schema[u'properties'][u'links'][u'additionalProperties']:
archivematica-mcp-server_1       |     {u'@comment': u'`link` is synonymous with MicroServiceChainLink and its related attributes including task manager configurations.',
archivematica-mcp-server_1       |      u'additionalProperties': False,
archivematica-mcp-server_1       |      u'properties': {u'config': {u'oneOf': [{u'$ref': u'#/definitions/link_model_StandardTaskFile'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_StandardTaskDir'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_ReplacementDic'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_ChainChoice'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_GetServiceOutput'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_GetServiceOutputPrompt'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_GetUnitVar'},
archivematica-mcp-server_1       |                                             {u'$ref': u'#/definitions/link_model_SetUnitVar'}]},
archivematica-mcp-server_1       |                      u'description': {u'$ref': u'#/definitions/translations'},
archivematica-mcp-server_1       |                      u'exit_codes': {u'additionalProperties': {u'properties': {u'job_status': {u'oneOf': [{u'$ref': u'#/definitions/job_status'},
archivematica-mcp-server_1       |                                                                                                           {u'type': u'null'}]},
archivematica-mcp-server_1       |                                                                                u'link_id': {u'oneOf': [{u'$ref': u'#/definitions/uuid'},
archivematica-mcp-server_1       |                                                                                                        {u'type': u'null'}]}},
archivematica-mcp-server_1       |                                                                u'type': u'object'},
archivematica-mcp-server_1       |                                      u'minProperties': 0,
archivematica-mcp-server_1       |                                      u'type': u'object'},
archivematica-mcp-server_1       |                      u'fallback_job_status': {u'oneOf': [{u'$ref': u'#/definitions/job_status'},
archivematica-mcp-server_1       |                                                          {u'type': u'null'}]},
archivematica-mcp-server_1       |                      u'fallback_link_id': {u'oneOf': [{u'$ref': u'#/definitions/uuid'},
archivematica-mcp-server_1       |                                                       {u'type': u'null'}]},
archivematica-mcp-server_1       |                      u'group': {u'$ref': u'#/definitions/translations'}},
archivematica-mcp-server_1       |      u'required': [u'config',
archivematica-mcp-server_1       |                    u'exit_codes',
archivematica-mcp-server_1       |                    u'fallback_job_status',
archivematica-mcp-server_1       |                    u'fallback_link_id',
archivematica-mcp-server_1       |                    u'description',
archivematica-mcp-server_1       |                    u'group'],
archivematica-mcp-server_1       |      u'type': u'object'}
archivematica-mcp-server_1       | 
archivematica-mcp-server_1       | On instance[u'links'][u'c2e6600d-cd26-42ed-bed5-95d41c06e37b']:
archivematica-mcp-server_1       |     {u'config_x': {u'@manager': u'linkTaskManagerDirectories',
archivematica-mcp-server_1       |                    u'@model': u'StandardTaskconfig_x',
archivematica-mcp-server_1       |                    u'arguments': u'"%SIPDirectory%" "%sharedPath%watchedDirectories/approveNormalization/." "%SIPUUID%" "%sharedPath%" "%SIPUUID%" "%sharedPath%"',
archivematica-mcp-server_1       |                    u'execute': u'moveSIP_v0.0',
archivematica-mcp-server_1       |                    u'filter_file_end': None,
archivematica-mcp-server_1       |                    u'filter_file_start': None,
archivematica-mcp-server_1       |                    u'filter_subdir': None,
archivematica-mcp-server_1       |                    u'stderr_file': None,
archivematica-mcp-server_1       |                    u'stdout_file': None},
archivematica-mcp-server_1       |      u'description': {u'en': u'Move to approve normalization directory',
archivematica-mcp-server_1       |                       u'es': u'Move to approve normalization directory',
archivematica-mcp-server_1       |                       u'fr': u'Move to approve normalization directory',
archivematica-mcp-server_1       |                       u'ja': u'Move to approve normalization directory',
archivematica-mcp-server_1       |                       u'pt_BR': u'Mover para aprovar o diret\xf3rio de normaliza\xe7\xe3o',
archivematica-mcp-server_1       |                       u'sv': u'Flytta till godk\xe4nn normaliserings-katalog'},
archivematica-mcp-server_1       |      u'exit_codes': {u'0': {u'job_status': u'Completed successfully',
archivematica-mcp-server_1       |                             u'link_id': None}},
archivematica-mcp-server_1       |      u'fallback_job_status': u'Failed',
archivematica-mcp-server_1       |      u'fallback_link_id': None,
archivematica-mcp-server_1       |      u'group': {u'en': u'Normalize',
archivematica-mcp-server_1       |                 u'es': u'Normalizar',
archivematica-mcp-server_1       |                 u'fr': u'Normaliser ',
archivematica-mcp-server_1       |                 u'ja': u'Normalize',
archivematica-mcp-server_1       |                 u'pt_BR': u'Normalizar',
archivematica-mcp-server_1       |                 u'sv': u'Normalisera'}}
compose_archivematica-mcp-server_1 exited with code 1
```

Reply:
:+1: